services/TipiVariable.py

```diff
@@ -0,0 +1,184 @@
+
+        
+        # Messages look like: (Tab-delimited)
+        #
+        # CHATTI     R   REMOTE_HOST     NULL        0   NULL        NULL
+        # CHATTI     W   REMOTE_PORT     9900        0   NULL        NULL
+        # CHATTI     W   MESSAGE         [myPasswd]  0   AUTHPASS    NULL
+        # CHATTI     W   MESSAGE         TESTING123  0   MESSAGE     [session_id]
+
+        message = self.tipi_io.receive()
+
+        # Now that we have message, let's parse it:
+        ti_message = message.split("\t")
+
+        caller_guid    = ti_message[0] if len(ti_message) >= 1 else ''  # Program's GUID
+        action         = ti_message[1] if len(ti_message) >= 2 else ''  # 'R', 'W', 'U', 'T'   read, write, transmit via UDP, transmit via TCP
+        var_key        = ti_message[2] if len(ti_message) >= 3 else ''  # Key
+        var_val        = ti_message[3] if len(ti_message) >= 4 else ''
+        queue          = ti_message[4] if len(ti_message) >= 5 else ''  # Allow values to queue for these variables. (only locally for now)
+        remote_var_key = ti_message[5] if len(ti_message) >= 6 else ''  # The name of the remote variable, if we're transmitting this var/val.
+        remote_id      = ti_message[6] if len(ti_message) >= 7 else ''  # Remote ID to present, if we're transmitting this var/val. We're going to use session_id for the remote_id
+
+
+        # Our response will be stored in "remote_var_key.RESP"
+        response = str(remote_var_key) + '.RESP' if (len(remote_var_key)) else ''
+
+        # Load our existing variables into a dict:
+        self.ti_vars = {}
+        
+        guid_file = '/home/tipi/tipi/RUN/' + caller_guid
+        
+        if os.path.isfile(str(guid_file)):
+            with open(str(guid_file), "r") as f:
+                for line in f:
+                    line = line.rstrip("\n\r")   # Strip newlines, but not whitespace (which is what rstrip() alone will do)
+                    data = line.split("\t")
+                    
+                    if len(data) > 1:
+                        self.ti_vars[data[0]] = data[1]
+
+            f.close
+
+        if action == 'W' or action == 'U' or action == 'T':   # Write!
+            self.ti_vars[response] = ''  # Blank out our old response
+        
+            var_val = var_val.rstrip()
+
+            self.ti_vars[str(var_key)] = str(var_val)
+                
+            self.store(caller_guid)   # Write our vars to our local file
+
+
+        if action == 'U':    #   TX via UDP    NOT COMPLETE! FIX!!!
+            sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+            server_address = (self.ti_vars['REMOTE_HOST'], int(self.ti_vars['REMOTE_PORT']))
+
+            sock.close()
+
+            self.store(caller_guid)   # Write our vars to our local file
+
+
+        if action == 'T':    #   TX via TCP
+            if 'REMOTE_HOST' not in self.ti_vars or 'REMOTE_PORT' not in self.ti_vars:
+                return bytearray("!ERROR!")
```

Comment:
I feel like the ! and _ characters around ERROR and FAIL are odd, and going to be hard to remember from a protocol client point of view... I like that the whole thing is a text based protocol. But I'd rather see a contract like, if the return message starts with ERROR:<some explanation> or FAIL:<some explanation>  then it is a TipiVariables error... 

Just something to think about.